### PR TITLE
Add interior lighting to stations

### PR DESCRIPTION
**Pull request by tatjam**
Makes ships that are inside stations (as defined on the model file using tags and a series of new parser commands) have increased ambient light and reduced direct lightning, fixing the issue of black ships on eclipsed stations (Issue #4399).

TO-DO:
- [x] Working interiors
- [x] Make all stations compatible with the code
- [ ] Adjustable lighting colors with some kind of station config?
- [x] Make it work with the ship cockpit and other bodies that may be found inside stations

Demo: https://www.youtube.com/watch?v=D56iUkvpOac

### CODE REVIEW

**Comment by fluffyfreak** [comment]
That's an interesting approach, another might be to gather the nearest 3 or 4 lights from within the station and apply them as light sources to the ship. However I'm not sure if "lights" like that are defined in the model or if that's something that could be done later.

**Comment by nozmajner** [comment]
There are no lights in the model. I can add them if needed. Not sure if Blender lights are exported to collada, with their settings. Or they could be tags.

**Comment by tatjam** [comment]
Love the idea, that way it's more in-line with the new PBR renderer idea. Guess adjusting ambient lighting is useful even then, if no dynamical env-maps are used.

Even if actual lights are used I guess the interior checking is important, to prevent outside ships from being lit. (Otherwise shadow casting is needed, or really thick exterior walls).

**Comment by nozmajner** [comment]
I'd suggest first doing the ambient light only, get it to merge, and then you can think about the next step, or if you want to do other features instead. No need to build a huge PR, especially when you are still getting your feet wet. This is a free-time passion project after all, avoid burnout and such!

**Comment by tatjam** [comment]
If I'm not mistaken all stations use either hub A o B so that should be done

**Comment by nozmajner** [comment]
Yes, all stations use those two hubs.

**Comment by impaktor** [comment]
@tatjam For testing big vs. small ships, I assume you're aware of Ctrl+i to change your ship model.

**Comment by tatjam** [comment]
Messed around a bit with various ships and a fixed fade value feels good. There's really not any ship that's so long its center is way outside the station before it visually is outside.

**Comment by impaktor** [comment]
@tatjam If you're feeling done, please rebase commits if/as needed, and remove draft status on PR

**Comment by tatjam** [comment]
Fine! The custom light feature could be implemented later on if needed.

**Comment by impaktor** [comment]
Going to need some review on this @Web-eWorks  or/and @fluffyfreak

**Comment by Web-eWorks** [comment]
Will do - not had a chance to get to this yet but it's been on my TODO list.

File: src/SpaceStation.cpp
Comment (fluffyfreak) [inline src/SpaceStation.cpp:917]: I think this should be the constant `Color::GRAY`

File: src/Camera.cpp
Comment (fluffyfreak) [inline src/Camera.cpp:602]: Replace with `Color::WHITE`

**Review by fluffyfreak — APPROVED** [review]
Aside from a couple of minor things which could be using the Color constants I think this is great work 👍

File: src/Camera.cpp
Comment (Web-eWorks) [inline src/Camera.cpp:565]: This function should perform early-out rejection if the distance between the body and the space station is less than `b->GetClipRadius() + ss->GetClipRadius()`.

File: src/scenegraph/Parser.cpp
Comment (Web-eWorks) [inline src/scenegraph/Parser.cpp:183]: Personally, I think `bound_cylinder` is a better choice of name here, but I'll leave it to your discretion whether thick_line is the technically correct name.

File: src/scenegraph/LoaderDefinitions.h
Comment (Web-eWorks) [inline src/scenegraph/LoaderDefinitions.h:73]: In general, I see this structure as unnecessarily future-proofed. Do you have any further plans to use `tags[2]` or `params[1]` for other bound types that you plan to implement immediately? If not, I'd recommend reducing the struct to only what's needed to express the minimum set of bounds currently in use.

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
This is certainly better than no lighting at all, and the implementation is fairly well-contained and easy to upgrade to shadow maps and punctual lights later down the line when we add support for them.

However, there are a few issues. The most notable one is that the new model bounds information will not be present when a model is converted to .SGM format when we ship a compiled build of the game, which will manifest as a complete lack of the new station lighting. The bounds information will need to be round-tripped through `src/scenegraph/BinaryConverter.cpp` in the `Save` and `Load` methods.

The second is that there is a worrying number of string comparisons and operations involved in computing interior lighting for stations. `Model::DistanceFromPointToBound` is called `N*M` times per frame, where N=number of ships on screen and M=number of stations within render distance. If a model contains more than one named bound type, there's a linear loop over bound entries by name. There is quite a bit of room for improvement here, starting with ignoring station-ship pairs with no overlap of visible cull radius. I'll leave some inline review comments where the behavior is non-optimal and presents scaling issues on lower-end hardware.

On the whole I'm quite happy with the PR and I do apologize that it took this long to review only to ask you to make major changes. I'd be happy to discuss further on the individual change requests if needed.

File: src/scenegraph/LoaderDefinitions.h
Comment (tatjam) [reply src/scenegraph/LoaderDefinitions.h:73]: I've got a feeling as most stations are going to be relatively linear, the cylinder / thick line thing is good enough. I'm going to change it to only support that, and if in the future other types are needed an union could be used.

Comment (tatjam) [reply src/scenegraph/LoaderDefinitions.h:73]: Actually using an union doesn't seem like a great idea if strings are used to store the tags, so probably just having different meanings for the tags. For example, an AABB can also be defined by an start and end tag, and the radius could be used for rounding.

AABBs and the cylinders can get you some really complex geometries!

File: src/scenegraph/Parser.cpp
Comment (tatjam) [reply src/scenegraph/Parser.cpp]: I've chosen "capsule" as a name instead, as this matches the typical capsule collider used in most game engines (cylinder with rounded endcaps)

Comment (Web-eWorks) [reply src/scenegraph/Parser.cpp]: Definitely a better choice - I would not have been able to tell that was the underlying shape from the existing nomenclature.

**Comment by Web-eWorks** [comment]
Just taking a first look over the code from my phone here - named struct field initializers are (AFAIK) a c++20 feature (which we don't currently target). Would recommend double-checking for that in the last few commits.

**Comment by tatjam** [comment]
Dangit! I've been lately working on a C++20 codebase and have gotten used to its quirks! Gonna change it back to simply setting the fields, which should be equivalent under any good compiler.

**Comment by tatjam** [comment]
How can I get the game to try loading the binary files instead of data?

**Comment by Web-eWorks** [comment]
You'll want to run the CMake `build-data` target through your editor or terminal of choice. That invokes the `modelcompiler` binary to build all models into `.sgm` files and leaves them in the install tree - you'll need to re-run the command when you modify the SGM format to rebuild the compiled models.

Keep in mind that you'll need to bump the SGM format version to add the bounds information to the binary stream that makes up an SGM file.

**Comment by tatjam** [comment]
Oh, but does the game automatically load the SGM files instead of the "raw" ones?

**Comment by tatjam** [comment]
Nevermind, it indeed does!

**Comment by tatjam** [comment]
I've left a "TODO", because the existing SGM code appeared to save the tags: https://github.com/pioneerspacesim/pioneer/blob/a7ce69ad3d87e1dee9a620a81b6124253217c1a9/src/scenegraph/BinaryConverter.cpp#L139
but they are not really used afterwards (or it appears so at a first glance, as they should be loaded from the model itself?)

**Review by Web-eWorks — APPROVED** [review]
With great apologies for the length of time this sat waiting for a review - this functions well and you've addressed my prior concerns very sufficiently.

If you're still present, I'd ask that you squash together the various "fix X" commits before merge; otherwise I can merge this PR as is.

**Comment by Web-eWorks** [comment]
> I've left a "TODO", because the existing SGM code appeared to save the tags:
> 
> https://github.com/pioneerspacesim/pioneer/blob/a7ce69ad3d87e1dee9a620a81b6124253217c1a9/src/scenegraph/BinaryConverter.cpp#L139
> 
> but they are not really used afterwards (or it appears so at a first glance, as they should be loaded from the model itself?)

I need to do a cleanup / fixup pass on the serialization code now that we've (finally) broken compatibility with file version 6. I wouldn't worry about this - I have a larger planned rewrite of how SGM files are internally structured.

**Comment by tatjam** [comment]
No problem! I've more or less cleaned up the commit history into somehow related blocks, although they could be further reduced.

**Comment by Web-eWorks** [comment]
Thank you @tatjam! Please let us know the name (handle, real name, etc.) you'd like to be listed as in AUTHORS.txt - this definitely meets the threshold to be added as a contributor. :smile:

(CC @impaktor for AUTHORS.txt / Changelog update)

**Comment by tatjam** [comment]
Btw I guess Issue #4399 is solved by this?

**Comment by fluffyfreak** [comment]
@tatjam I just need to echo what others have said, thankyou so much for doing this! 
It was one of the very annoying issues that needed to be fixed and I think you've done a really good job here.

**Pull request merged**
